File: compare/test-case-generator.js
var fs = require('fs');
var path = require('path');

var MANUAL_CASES = [
  { origin: '三桥', dest: '半坡', desc: '同线直达(1号线)' },
  { origin: '小寨', dest: '钟楼', desc: '1次换乘大站' },
  { origin: '丈八北路', dest: '万寿路', desc: '1次换乘多选' },
  { origin: '保税区', dest: '航天新城', desc: '2次换乘长距离' },
  { origin: '杨官寨', dest: '秦陵西', desc: '跨远郊超长路线' },
  { origin: '北大街', dest: '钟楼', desc: '近距离相邻站' },
  { origin: '后卫寨', dest: '纺织城', desc: '1号线全程' },
  { origin: '草滩', dest: '常宁宫', desc: '2号线全程' },
  { origin: '鱼化寨', dest: '保税区', desc: '3号线长距离' },
  { origin: '创新港', dest: '纺织城', desc: '5号线全程' },
];

function haversineMeters(lat1, lon1, lat2, lon2) {
  var R = 6371000;
  var dLat = (lat2 - lat1) * Math.PI / 180;
  var dLon = (lon2 - lon1) * Math.PI / 180;
  var a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
    Math.sin(dLon / 2) * Math.sin(dLon / 2);
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

function generate(stationsData, config) {
  var stationMap = {};
  stationsData.forEach(function (s) {
    stationMap[s.name] = s;
  });

  var cases = [];
  var id = 1;

  MANUAL_CASES.forEach(function (mc) {
    var oStation = stationMap[mc.origin];
    var dStation = stationMap[mc.dest];
    if (!oStation || !dStation) return;
    cases.push({
      id: 'T' + String(id++).padStart(2, '0'),
      type: 'manual',
      desc: mc.desc,
      origin: mc.origin,
      dest: mc.dest,
      originCoord: [oStation.lon, oStation.lat],
      destCoord: [dStation.lon, dStation.lat],
    });
  });

  var names = stationsData.map(function (s) { return s.name; });
  var used = {};
  cases.forEach(function (c) {
    used[c.origin + '→' + c.dest] = true;
    used[c.dest + '→' + c.origin] = true;
  });

  var maxAttempts = config.RANDOM_CASE_COUNT * 10;
  var attempts = 0;
  while (cases.length < 10 + config.RANDOM_CASE_COUNT && attempts < maxAttempts) {
    attempts++;
    var oi = Math.floor(Math.random() * names.length);
    var di = Math.floor(Math.random() * names.length);
    if (oi === di) continue;

    var oName = names[oi];
    var dName = names[di];
    var key = oName + '→' + dName;
    if (used[key]) continue;

    var oS = stationMap[oName];
    var dS = stationMap[dName];
    var dist = haversineMeters(oS.lat, oS.lon, dS.lat, dS.lon);
    if (dist < config.MIN_DISTANCE_M) continue;

    used[key] = true;
    used[dName + '→' + oName] = true;

    cases.push({
      id: 'T' + String(id++).padStart(2, '0'),
      type: 'random',
      origin: oName,
      dest: dName,
      originCoord: [oS.lon, oS.lat],
      destCoord: [dS.lon, dS.lat],
    });
  }

  return cases;
}

module.exports = { generate: generate };

File: compare/test_case_generator.py
import math
import random


MANUAL_CASES = [
    {"origin": "三桥", "dest": "半坡", "desc": "same line direct"},
    {"origin": "小寨", "dest": "钟楼", "desc": "single transfer core section"},
    {"origin": "丈八北路", "dest": "万寿路", "desc": "single transfer with options"},
    {"origin": "保税区", "dest": "航天新城", "desc": "long multi-transfer case"},
    {"origin": "杨官寨", "dest": "秦陵西", "desc": "outer suburban long route"},
    {"origin": "北大街", "dest": "钟楼", "desc": "short adjacent route"},
    {"origin": "后卫寨", "dest": "纺织城", "desc": "line 1 full trip"},
    {"origin": "草滩", "dest": "常宁宫", "desc": "line 2 full trip"},
    {"origin": "鱼化寨", "dest": "保税区", "desc": "line 3 long route"},
    {"origin": "创新港", "dest": "纺织城", "desc": "line 5 full trip"},
]


def haversine_meters(lat1, lon1, lat2, lon2):
    radius = 6371000
    d_lat = math.radians(lat2 - lat1)
    d_lon = math.radians(lon2 - lon1)
    value = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.sin(d_lon / 2) ** 2
    )
    return radius * 2 * math.atan2(math.sqrt(value), math.sqrt(1 - value))


def generate(stations, config):
    station_map = {station["name"]: station for station in stations}
    cases = []
    used = set()

    for index, manual_case in enumerate(MANUAL_CASES, start=1):
        origin_station = station_map.get(manual_case["origin"])
        dest_station = station_map.get(manual_case["dest"])
        if not origin_station or not dest_station:
            continue
        cases.append(
            {
                "id": f"T{index:02d}",
                "type": "manual",
                "desc": manual_case["desc"],
                "origin": manual_case["origin"],
                "dest": manual_case["dest"],
                "origin_coord": [origin_station["lon"], origin_station["lat"]],
                "dest_coord": [dest_station["lon"], dest_station["lat"]],
            }
        )
        used.add((manual_case["origin"], manual_case["dest"]))
        used.add((manual_case["dest"], manual_case["origin"]))

    names = [station["name"] for station in stations]
    target_count = len(cases) + config.RANDOM_CASE_COUNT
    attempts = 0
    max_attempts = config.RANDOM_CASE_COUNT * 10
    next_index = len(cases) + 1

    while len(cases) < target_count and attempts < max_attempts:
        attempts += 1
        origin_name, dest_name = random.sample(names, 2)
        if (origin_name, dest_name) in used:
            continue

        origin_station = station_map[origin_name]
        dest_station = station_map[dest_name]
        distance = haversine_meters(
            origin_station["lat"],
            origin_station["lon"],
            dest_station["lat"],
            dest_station["lon"],
        )
        if distance < config.MIN_DISTANCE_M:
            continue

        used.add((origin_name, dest_name))
        used.add((dest_name, origin_name))
        cases.append(
            {
                "id": f"T{next_index:02d}",
                "type": "random",
                "origin": origin_name,
                "dest": dest_name,
                "origin_coord": [origin_station["lon"], origin_station["lat"]],
                "dest_coord": [dest_station["lon"], dest_station["lat"]],
            }
        )
        next_index += 1

    return cases
